Create an AlignmentFile object for “merged-tumor.bam” from Public files gallery

In [1]:
import pysam
alignment_file = pysam.AlignmentFile("merged-tumor.bam", "rb")

In [2]:
alignment_array = []
for read in alignment_file:
    alignment_array.append(read)

Take the first read from the AlignmentFile

In [3]:
print(alignment_array[0])

C0HVYACXX120402:7:1207:5722:57044	1187	20	9483248	27	76M	20	9483381	76	TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG	array('B', [28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])	[('XA', 'GL000217.1,-110754,76M,1;'), ('BD', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('MD', '76'), ('RG', '1'), ('BI', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('NM', 0), ('MQ', 27), ('AS', 76), ('XS', 71)]


Inspect the fields in the AlignedSegment

In [4]:
print("Query name:"); display(alignment_array[0].query_name)
print("Query sequence:"); display(alignment_array[0].query_sequence)
print("Flag:"); display(alignment_array[0].flag)
print("Reference id:"); display(alignment_array[0].reference_id)
print("Reference start:"); display(alignment_array[0].reference_start)
print("Mapping quality:"); display(alignment_array[0].mapping_quality)
print("CIGAR:"); display(alignment_array[0].cigar)
print("Next reference id:"); display(alignment_array[0].next_reference_id)
print("Next reference start:"); display(alignment_array[0].next_reference_start)
print("Template length:"); display(alignment_array[0].template_length)
print("Query qualities:"); display(alignment_array[0].query_qualities)
print("Tags:"); display(alignment_array[0].tags)

Query name:


'C0HVYACXX120402:7:1207:5722:57044'

Query sequence:


'TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG'

Flag:


1187

Reference id:


20

Reference start:


9483248

Mapping quality:


27

CIGAR:


[(0, 76)]

Next reference id:


20

Next reference start:


9483381

Template length:


209

Query qualities:


array('B', [28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])

Tags:


[('XA', 'GL000217.1,-110754,76M,1;'),
 ('BD',
  'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'),
 ('MD', '76'),
 ('RG', '1'),
 ('BI',
  'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'),
 ('NM', 0),
 ('MQ', 27),
 ('AS', 76),
 ('XS', 71)]

Inspect the flag field

In [5]:
display(alignment_array[0].flag)

1187

1187: (fetched from https://broadinstitute.github.io/picard/explain-flags.html)
<br>    read paired (0x1)
<br>    read mapped in proper pair (0x2)
<br>    mate reverse strand (0x20)
<br>    second in pair (0x80)
<br>    read is PCR or optical duplicate (0x400)

Calculate how many unmapped reads are in the file?

In [6]:
mapped = 0
unmapped = 0
cnt = 0
for item in alignment_array:
    if item.is_unmapped:
        unmapped += 1
    else:
        mapped += 1
    cnt += 1
display("Mapped   = " + str(mapped))
display("Unmapped = " + str(unmapped))
display("Total    = " + str(cnt))    

'Mapped   = 2903864'

'Unmapped = 17765'

'Total    = 2921629'

Calculate total number of reads

In [7]:
len(alignment_array)

2921629

Calculate the number of reads with mapping quality 0

In [8]:
cnt = 0
for item in alignment_array:
    if item.mapping_quality == 0:
        cnt += 1
cnt

126628

Calculate the average mapping quality for all the reads

In [9]:
num = 0
quality = 0
for item in alignment_array:
    quality += item.mapping_quality
    num += 1
display("Average mapping quality = " + str(quality/num))

'Average mapping quality = 55.91379158681681'

Calculate the average mapping quality if reads with 0 map quality are filtered out

In [10]:
num = 0
quality = 0
for item in alignment_array:
    if item.mapping_quality != 0: 
        quality += item.mapping_quality
        num += 1
display("Num without 0 mapping quality = " + str(num)) 
display("Num with 0 mapping quality = " + str(len(alignment_array)-num))        
display("Average mapping quality = " + str(quality/num))

'Num without 0 mapping quality = 2795001'

'Num with 0 mapping quality = 126628'

'Average mapping quality = 58.446975510921106'